File: scripts/ListCartoriosRestransmitirOe/listCartoriosRestransmitirOe.py
#!/usr/bin/env python3
"""Lista cartórios permitidos para retransmissão (ListCartoriosRestransmitirOE) no webservice Ofícios da ONR."""

from __future__ import annotations

import json
import sys
from pathlib import Path

from dotenv import load_dotenv

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))

from lib.onr_json import to_json_safe  # noqa: E402
from lib.onr_oficios import (  # noqa: E402
    hash_error_hint,
    load_login_config,
    load_oficios_soap_config,
    load_serventia_chave,
    resolve_auth_hash,
)
from lib.onr_soap import call_operation_from_cfg  # noqa: E402
from lib.onr_zeep_serialize import serialize_zeep_list  # noqa: E402

FIELD_ORDER = ("Hash",)
CARTORIO_ITEM = "ListRIRetransOE_WSResp"


def _business_error_hint(codigo: int) -> str | None:
    hints = {
        51: "Apenas usuários de cartórios são permitidos.",
        52: "Não foi possível obter os dados do cartório do usuário.",
        53: "Não foi possível obter os cartórios.",
    }
    return hints.get(codigo)


def load_config() -> dict:
    load_dotenv(ROOT / ".env")
    return {
        "chave": load_serventia_chave(),
        "login_cfg": load_login_config(),
        **load_oficios_soap_config(),
    }


def build_request(hash_value: str) -> dict:
    return {key: hash_value for key in FIELD_ORDER}


def build_response(result) -> dict:
    return {
        "RETORNO": result.RETORNO,
        "CODIGOERRO": result.CODIGOERRO,
        "ERRODESCRICAO": result.ERRODESCRICAO,
        "Cartorios": serialize_zeep_list(result.Cartorios, CARTORIO_ITEM),
    }


def main() -> int:
    cfg = load_config()
    hash_value = resolve_auth_hash(cfg["chave"], cfg["login_cfg"])
    o_request = build_request(hash_value)

    print("=== Parâmetros ListCartoriosRestransmitirOE ===")
    print(json.dumps(o_request, ensure_ascii=False, indent=2))
    print(f"\nEndpoint: {cfg['endpoint']}")

    result = call_operation_from_cfg(cfg, "ListCartoriosRestransmitirOE", o_request)
    response = to_json_safe(build_response(result))

    print("\n=== Resposta ===")
    print(json.dumps(response, ensure_ascii=False, indent=2))

    if not response.get("RETORNO"):
        codigo = response.get("CODIGOERRO", "?")
        descricao = response.get("ERRODESCRICAO", "")
        print(
            f"\nListCartoriosRestransmitirOE falhou: [{codigo}] {descricao}",
            file=sys.stderr,
        )
        if isinstance(codigo, int):
            hint = _business_error_hint(codigo) or hash_error_hint(codigo)
            if hint:
                print(hint, file=sys.stderr)
        return 1

    cartorios = response.get("Cartorios") or []
    print(f"\nOK — {len(cartorios)} cartório(s) permitido(s) para retransmissão.")
    for row in cartorios[:15]:
        print(
            f"  IDCartorio={row.get('IDCartorio')}: {row.get('Cartorio', '—')}"
        )
    if len(cartorios) > 15:
        print(f"  ... e mais {len(cartorios) - 15}")
    print(
        "\nUse IDCartorio em OFICIOS_SET_PEDIDO_RETRANSMITIDO_ID_CARTORIO "
        "(SetPedidoRetransmitidoOE)."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/ListCartoriosRestransmitirOe/listCartoriosRestransmitirOe.js
/**
 * Lista cartórios permitidos para retransmissão (ListCartoriosRestransmitirOE) no webservice Ofícios da ONR.
 */
import fs from "node:fs";
import soap from "soap";
import { resolvePath } from "../../lib/onr_env.js";
import {
  hashErrorHint,
  loadLoginConfig,
  loadOficiosSoapConfig,
  loadServentiaChave,
  resolveAuthHash,
} from "../../lib/onr_oficios.js";

const CARTORIO_ITEM = "ListRIRetransOE_WSResp";
const FIELD_ORDER = ["Hash"];

function loadConfig() {
  return {
    chave: loadServentiaChave(),
    loginCfg: loadLoginConfig(),
    ...loadOficiosSoapConfig(),
  };
}

function buildRequest(hashValue) {
  return Object.fromEntries(FIELD_ORDER.map((key) => [key, hashValue]));
}

function serializeCartorios(cartorios) {
  if (!cartorios) return [];
  const items = cartorios[CARTORIO_ITEM];
  if (!items) return [];
  return Array.isArray(items) ? items : [items];
}

function normalizeResponse(result) {
  return {
    RETORNO: result.RETORNO,
    CODIGOERRO: result.CODIGOERRO,
    ERRODESCRICAO: result.ERRODESCRICAO,
    Cartorios: serializeCartorios(result.Cartorios),
  };
}

function businessErrorHint(codigo) {
  const hints = {
    51: "Apenas usuários de cartórios são permitidos.",
    52: "Não foi possível obter os dados do cartório do usuário.",
    53: "Não foi possível obter os cartórios.",
  };
  return hints[codigo] ?? null;
}

async function listCartoriosRestransmitirOe(cfg, oRequest) {
  const wsdlPath = resolvePath(cfg.wsdlPath);
  if (!fs.existsSync(wsdlPath)) {
    throw new Error(`WSDL não encontrado: ${wsdlPath}`);
  }

  const client = await soap.createClientAsync(wsdlPath, {
    endpoint: cfg.endpoint,
    forceSoap12Headers: false,
    wsdl_options: { timeout: 60000 },
  });

  let response;
  try {
    [response] = await client.ListCartoriosRestransmitirOEAsync({ oRequest });
  } catch (err) {
    const status = err?.response?.status;
    const body =
      typeof err?.body === "string"
        ? err.body
        : typeof err?.response?.data === "string"
          ? err.response.data
          : null;

    if (status === 503 || (body && !body.trimStart().startsWith("<"))) {
      throw new Error(
        `Servidor ONR indisponível (HTTP ${status ?? "?"}): ${body ?? err.message}`
      );
    }
    throw err;
  }

  const result =
    response?.ListCartoriosRestransmitirOEResult ??
    response?.ListCartoriosRestransmitirOEResponse?.ListCartoriosRestransmitirOEResult ??
    response;

  return normalizeResponse(result);
}

async function main() {
  const cfg = loadConfig();
  const hash = await resolveAuthHash(cfg.chave, cfg.loginCfg);
  const oRequest = buildRequest(hash);

  console.log("=== Parâmetros ListCartoriosRestransmitirOE ===");
  console.log(JSON.stringify(oRequest, null, 2));
  console.log(`\nEndpoint: ${cfg.endpoint}`);

  const response = await listCartoriosRestransmitirOe(cfg, oRequest);
  console.log("\n=== Resposta ===");
  console.log(JSON.stringify(response, null, 2));

  if (!response.RETORNO) {
    console.error(
      `\nListCartoriosRestransmitirOE falhou: [${response.CODIGOERRO}] ${response.ERRODESCRICAO}`
    );
    const hint =
      businessErrorHint(response.CODIGOERRO) ?? hashErrorHint(response.CODIGOERRO);
    if (hint) console.error(hint);
    process.exit(1);
  }

  const cartorios = response.Cartorios ?? [];
  console.log(`\nOK — ${cartorios.length} cartório(s) permitido(s) para retransmissão.`);
  for (const row of cartorios.slice(0, 15)) {
    console.log(`  IDCartorio=${row.IDCartorio}: ${row.Cartorio ?? "—"}`);
  }
  if (cartorios.length > 15) {
    console.log(`  ... e mais ${cartorios.length - 15}`);
  }
  console.log(
    "\nUse IDCartorio em OFICIOS_SET_PEDIDO_RETRANSMITIDO_ID_CARTORIO (SetPedidoRetransmitidoOE)."
  );
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});
